ko/resolve/resolve.go

```diff
@@ -0,0 +1,163 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resolve
+
+import (
+	"bytes"
+	"io"
+	"sync"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/google/go-containerregistry/ko/build"
+	"github.com/google/go-containerregistry/ko/publish"
+)
+
+// ImageReferences resolves supported references to images within the input yaml
+// to published image digests.
+func ImageReferences(input []byte, builder build.Interface, publisher publish.Interface) ([]byte, error) {
+	// The loop is to support multi-document yaml files.
+	// This is handled by using a yaml.Decoder and reading objects until io.EOF, see:
+	// https://github.com/go-yaml/yaml/pull/301/files#diff-e51f89b39d976550831afb510a926be0R123
+	decoder := yaml.NewDecoder(bytes.NewBuffer(input))
+	buf := bytes.NewBuffer(nil)
+	encoder := yaml.NewEncoder(buf)
+	for {
+		var obj interface{}
+		if err := decoder.Decode(&obj); err != nil {
+			if err == io.EOF {
+				return buf.Bytes(), nil
+			}
+			return nil, err
+		}
+
+		// Recursively walk input, building and publishing supported references.
+		// TODO(mattmoor): It may be worth considering gathering the supported references in
+		// a first pass, performing the builds and pushes, and then performing the
+		// substitutions in a second pass.  This would enable us to eliminate the goroutine
+		// logic embedded in the recursive walk, naturally eliminate redundant build / publish
+		// for the same reference, and potentially enable us to create a single larger build
+		// (not clear if/how much this is possible in go build).
+		obj2, err := replaceRecursive(obj, func(ref string) (string, error) {
+			if !builder.IsSupportedReference(ref) {
+				return ref, nil
+			}
+			img, err := builder.Build(ref)
+			if err != nil {
+				return "", err
+			}
+			digest, err := publisher.Publish(img, ref)
+			if err != nil {
+				return "", err
+			}
+			return digest.String(), nil
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		if err := encoder.Encode(obj2); err != nil {
+			return nil, err
+		}
+	}
+}
+
+type replaceString func(string) (string, error)
+
+// replaceRecursive walks the provided untyped object recursively by switching
+// on the type of the object at each level. It supports walking through the
+// keys and values of maps, and the elements of an array. When a leaf of type
+// string is encountered, this will call the provided replaceString function on
+// it. This function does not support walking through struct types, but also
+// should not need to as the input is expected to be the result of parsing yaml
+// or json into an interface{}, which should only produce primitives, maps and
+// arrays. This function will return a copy of the object rebuilt by the walk
+// with the replacements made.
+func replaceRecursive(obj interface{}, rs replaceString) (interface{}, error) {
+	switch typed := obj.(type) {
+	case map[interface{}]interface{}:
+		sm := sync.Map{}
+		wg := sync.WaitGroup{}
```

Comment:
supernit: I typically see this as `var wg sync.WaitGroup` (and `var sm sync.Map`)

Reply:
Done.